ios/uPortMobile/Classes/EthereumSigner.m

```diff
@@ -0,0 +1,203 @@
+  NSDictionary *sig = genericSignature(keypair, hash, YES);
+  if (!sig) return NULL;
+  NSData *rData = (NSData *)sig[@"r"];
+  NSData *sData = (NSData *)sig[@"s"];
+  BN_ULONG base = 0x1b; // pre-EIP155
+  if (chainId) {
+    BIGNUM *v = BN_new(); BN_bin2bn(chainId.bytes, chainId.length, v);
+    // TODO support longer chainIDs
+    base = BN_get_word(v) * 2 + 35;
+    BN_clear_free(v);
+  }
+  
+  NSDictionary *signatureDictionary = @{ @"v": @(base + [sig[@"recoveryParam"] intValue]),
+                                         @"r": [rData base64EncodedStringWithOptions:0],
+                                         @"s":[sData base64EncodedStringWithOptions:0]};
+  return signatureDictionary;
+}
+
+
+NSDictionary *genericSignature(BTCKey *keypair, NSData *hash, BOOL lowS) {
+    BTCBigNumber* privkeyBN = [[BTCBigNumber alloc] initWithUnsignedBigEndian:[keypair privateKey]];
+    BTCBigNumber* n = [BTCCurvePoint curveOrder];
+  
+    NSMutableData* kdata = [keypair signatureNonceForHash:hash];
+    BTCMutableBigNumber* k = [[BTCMutableBigNumber alloc] initWithUnsignedBigEndian:kdata];
+    [k mod:n]; // make sure k belongs to [0, n - 1]
+    
+    if (BN_cmp(k.BIGNUM, [BTCBigNumber zero].BIGNUM) <= 0 && BN_cmp(k.BIGNUM, n.BIGNUM) >= 0) return NULL;
+
+    BTCDataClear(kdata);
+  
+    BTCCurvePoint* K = [[BTCCurvePoint generator] multiply:k];
+    // K was infinity and does not work, redo it
+    if ([K isInfinity])  return NULL;
+    BTCMutableBigNumber* Kx = [K.x mutableCopy];
+    [Kx mod: n];
+    
+    BTCBigNumber* hashBN = [[BTCBigNumber alloc] initWithUnsignedBigEndian:hash];
+  
+    // Compute s = (k^-1)*(h + Kx*privkey)
+  
+    BTCMutableBigNumber* signatureBN = [[[[privkeyBN mutableCopy] multiply:Kx mod:n] add:hashBN mod: n] multiply:[k inverseMod:n] mod:n];
+
+    // Neither r nor s can be zero
+    if ([Kx isZero] || [signatureBN isZero]) return NULL;
+    BTCBigNumber *r = [Kx copy];
+    uint8_t recoveryParam = BN_is_odd(K.y.BIGNUM) ? 1 : 0;
+    
+    [privkeyBN clear];
+    [k clear];
+    [hashBN clear];
+    [K clear];
+    [Kx clear];
+
+    BTCBigNumber *twiceS = [[signatureBN mutableCopy] add: signatureBN];
+    
+    BTCBigNumber *s;
+    if (lowS && BN_cmp(twiceS.BIGNUM, n.BIGNUM) > 0) {
+      // enforce low S values, by negating the value (modulo the order) if above order/2.
+      s = [[n mutableCopy] subtract: signatureBN]; // Maybe this should be swapped
+      recoveryParam ^= 1;
+    } else {
+      s = [signatureBN copy];
+    }
+
+    [twiceS clear];
+    [signatureBN clear];
+
+    NSData* rData = [r unsignedBigEndian];
+    NSData* sData = [s unsignedBigEndian];
+    [r clear];
+    [s clear];
+    
+    return @{
+             @"r": rData,
+             @"s": sData,
+             @"recoveryParam": @(recoveryParam)
+             };
+}
+
+NSData *simpleSignature(BTCKey *keypair, NSData *hash) {
+  NSDictionary *sig = genericSignature(keypair, hash, NO);
+  if (!sig) return NULL;
```

Comment:
Although it's correct in C, what you're actually testing if `sig` is `nil`.  So to keep close to that meaning, I think it's 'better' to write `if (sig == nil)`